java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/Server.java

```diff
@@ -869,11 +871,42 @@ public CompletableFuture<InitializeResult> initialize(InitializeParams init) {
                 }
             }
             CompletableFuture<Project[]> prjs = workspaceProjects;
-            SERVER_INIT_RP.post(() -> asyncOpenSelectedProjects0(prjs, projectCandidates, true, true));
+            SERVER_INIT_RP.post(() -> {
+                List<FileObject> additionalCandidates = new ArrayList<>();
+                AtomicBoolean cancel = new AtomicBoolean();
+                ProgressHandle h = ProgressHandle.createHandle("Collecting workspace projects...", () -> {
+                    cancel.set(true);
+                    return true;
+                });
+                h.start();
+                try {
+                    for (FileObject candidate : projectCandidates) {
```

Comment:
I think this body loop is much the same as `collectProjectCandidates`  body loop; maybe worth joining ?

Reply:
Code rewritten.